intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+
+######################################################################
+# The ``query = query.refine_names(..., 'T', 'D')`` serves as enforcable documentation
+# and lifts input dimensions to being named. It checks that the last two dimensions
+# can be refined to `['T', 'D']`, preventing potentially silent or confusing size
+# mismatch errors later down the line.
+#
+# **(II)  Manipulating dimensions in ``prepare_head``**
+
+# (II)
+def prepare_head(tensor):
+    tensor = tensor.refine_names(..., 'T', 'D')
+    return (tensor.unflatten('D', [('H', n_heads), ('D_head', dim_per_head)])
+                  .align_to(..., 'H', 'T', 'D_head'))
+
+######################################################################
+# The first thing to note is how the code clearly states the input and
+# output dimensions: the input tensor must end with the `T` and `D` dims
+# and the output tensor ends in `H`, `T`, and `D_head` dims.
+#
+# The second thing to note is how clearly the code describes what is going on.
+# prepare_head takes the key, query, and value and splits the embedding dim into
+# multiple heads, finally rearranging the dim order to be `[..., 'H', 'T', 'D_head']`.
+# ParlAI implements prepare_head as the following, using `view` and `transpose`
+# operations:
+#
+# **(III) Explicit broadcasting by names**
+
+def prepare_head(tensor):
+    # input is [batch_size, seq_len, n_heads * dim_per_head]
+    # output is [batch_size * n_heads, seq_len, dim_per_head]
+    batch_size, seq_len, _ = tensor.size()
+    tensor = tensor.view(batch_size, tensor.size(1), n_heads, dim_per_head)
+    tensor = (
+        tensor.transpose(1, 2)
+        .contiguous()
+        .view(batch_size * n_heads, seq_len, dim_per_head)
+    )
+    return tensor
+
+######################################################################
+# Our named tensor variant uses ops that, though more verbose, also have
+# more semantic meaning than `view` and `transpose` and include enforcable
+# documentation in the form of names.
+#
+# **(III) Explicit broadcasting by names**
+
+def ignore():
+    # (III)
+    # Named tensors doesn't support == yet; the following is a workaround.
+    attn_mask = (mask.renamed(None) == 0).refine_names(*mask.names)
+
+    # recall that we had dot_prod.refine_names(..., 'H', 'T', 'T_key')
+    attn_mask = attn_mask.align_as(dot_prod)
+
+    dot_prod.masked_fill_(attn_mask, -float(1e20))
+
+######################################################################
+# ``mask`` usually has dims ``[N, T]`` (in the case of self attention) or
+# ``[N, T, T_key]`` (in the case of encoder attention) while ``dot_prod``
+# has dims ``[N, H, T, T_key]``. To make ``mask`` broadcast correctly with
+# ``dot_prod``, we would usually `unsqueeze` dims `1` and `-1` in the case of self
+# attention or `unsqueeze` dim `1` in the case of encoder attention. Using
+# named tensors, we can simply align the two tensors and stop worrying about

```

Comment:
Sentence repeated twice

Reply:
Which one? I am blind apparently and don't see it.